### DISCUSSION TITLE
📢 Call for Feedback for WinUI Community Calls

### DISCUSSION THREAD

**Discussion by gabbybilka** [microsoft/microsoft-ui-xaml#7879] [General]
As we kick off Year 3+ (🎉!) of the WinUI community call I wanted to invite you, our community, to reflect and leave comments on what you would like to see in the monthly call. Thank you in advance for your feedback!

If you're unfamiliar with the WinUI community call, it is a call among the WinUI team that is live-streamed onto YouTube. We present on new updates, share information, welcome guests, and answer your questions. Check out this month's call issue for a recent example: #6815 

A few questions we're curious about:
- What should we continue to do/do more of?
- What should we stop doing/do less of?
- What is missing from community calls?
- How are we doing balancing WinUI & Windows App SDK information?
- Should we make changes to the format of the call? E.g. more formal/informal, more time for a specific content area, hosted on a different platform?

At a general level, we're also interested in what kinds of demos & spotlights you most appreciate. A few top of mind ideas are:
- Brown bags/feature walkthroughs with members of the WinUI team
- Intro to Windows App SDK & WinUI 3 series
- Demos of applications using WinUI & their development journey

We appreciate all feedback that you send our way! Unfortunately, not all feedback shared will be able to be implemented, but all will be considered to help shape the community call to be what **you** want it to be 🚀

**Comment by kmgallahan** [comment]
> What is missing from community calls?

A list of deep dive topics populated by the WinUI team & community could be provided and voted on. The winner(s) could then get a companion blog post and be discussed on the call. 

Most of the code isn't open source, so talking about the closed source code and sharing diagrams/details and explaining what makes working with it complicated (I imagine) would be interesting. 

The blog post could be released before the day of the call so the community could be ready with follow-up questions.

Examples:

* There used to be somewhat frequent blog posts like [this](https://blogs.windows.com/windowsdeveloper/2016/08/26/interop-between-xaml-and-the-visual-layer/#uljyfkZhh2fmTOEP.97) or [this](https://github.com/microsoft/WindowsCompositionSamples/wiki/XAML-Composition-Interop-Behavior-Changes), or [a follow up to this](https://jeremiahmorrill.wordpress.com/2011/02/14/a-critical-deep-dive-into-the-wpf-rendering-system/). I'd like to know how rendering has changed from UWP XAML to WinUI 3. This would probably help shed light on why adding support for acrylic & mica isn't trivial.

* What could be a process for prioritizing servicing of existing controls, such as [TextBox with 86 open issues](https://github.com/microsoft/microsoft-ui-xaml/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-TextBox). I imagine one team member could be assigned to get really familiar with the relevant code and knock out a bunch of issues over the course of 1-2 weeks. But deciding which controls to focus on, issue prioritization, bugs vs. features would all be a puzzle. Presenting a plan / outline to tackle servicing of a control (or other area) before a call to get final feedback and suggestions would be great. Then discuss the plan on the call.

**Comment by wbokkers** [comment]
Please stop ignoring feedback given during community calls on major issues, like:

- layout cycle crashed due to a scrollviewer bug: https://github.com/microsoft/microsoft-ui-xaml/issues/6218
- drag/drop crashes: https://github.com/microsoft/microsoft-ui-xaml/issues/4574
- Recently I found that a lot of our users run into an issue with the clipboard: https://github.com/microsoft/microsoft-ui-xaml/issues/6754
- Random WinRT crashes. WinRT exceptions often fail fast, without any indication of why it is happening.
- Out of memory exceptions, probably caused by a memory leak in NavigationView: https://github.com/microsoft/microsoft-ui-xaml/issues/5978

**Comment by wbokkers** [comment]
The servicing release is very disappointing. None of the issues I listed are addressed afaik. And they are major issues.

**Comment by kmgallahan** [comment]
@wbokkers unclear how complaining about open bugs relates to the subject of this issue.

**Comment by wbokkers** [comment]
@kmgallahan unclear how these issues are not related? These issues are open for months. Occurring in a big app. And we really, really want them to go away. I tried every channel to bring them to the attention of Microsoft.

**Comment by kmgallahan** [comment]
@wbokkers This issue is asking about how to improve the community calls. I don't see how providing a list of open bugs translates to a change you'd like for future community calls.

> we really, really want them to go away. I tried every channel to bring them to the attention of Microsoft.

Everyone wants their bugs to get fixed. That doesn't mean everyone should list their bugs here.

**Comment by wbokkers** [comment]
@kmgallahan I am not sure why you are the one to decide what is important to another member of the community. 
To me improving community calls is start listening to all members of the community. This is not about my issues per se, it is about not taking issues seriously. In the previous Community Call, as you might remember, there was a lot of talk about this.

**Comment by mrlacey** [comment]
I'd like to see this community call take on some of the things we see in community calls from other Microsoft teams.
Things like:
- progress on what has happened with the project since the last call.
- Notable community activity, blog posts, etc.
- occasional deep dives with members of the team on specific areas

**Comment by kmgallahan** [comment]
@wbokkers again, this issue is about improving **the community calls themselves**, not WinUI. Fixing misc bugs isn't a change to the community calls.

**Comment by MEK3DK** [comment]
> @wbokkers again, this issue is about improving **the community calls themselves**, not WinUI. Fixing misc bugs isn't a change to the community calls.

@kmgallahan I agree with you, but the problem atm is that there is no feedback from the team on issues in this forum. To many this is very frustrating and community calls seems to be the only place to raise issues/get a response.

**Comment by wbokkers** [comment]
@kmgallahan To me, improving the community calls themselves is focusing more on what the community, making real world apps and relying on WinUI, is coming across using WinUI. It's a **WinUI** Community Call after all. This kind of conceptual blah blah about the purpose of this post is a bit too much.... What is the best place to show my frustration with some change Microsoft is listening? During Community Calls themselves, my feedback systematically gets ignored. So that obviously has to change.

**Comment by kmgallahan** [comment]
@MEK3DK Yes, and the normal community call announcements are a great place to ask for feedback. Here though:

MS: "How can we improve future calls?"
Reply: "Here are some bugs"
MS: "???"

**Comment by wbokkers** [comment]
MS: "How can we improve future calls?"
Reply: "Here's a list of bugs I repeatedly bring up during Community Calls: don't ignore me"
MS: Ah, I see. Maybe we should show progress on issues and give insight in how we prioritize those issues.
Reply: "Yeah, exactly, it didn't cross my mind to formulate it in this better way, but thank you for understanding what's going on"

**Comment by MrDeej** [comment]
Great that multi-window support is included in 1.01, looking forward to trying it out!

**Comment by myokeeh** [comment]
Just about everything I've used for my current apps broke or are in limbo. Can we talk about progress on those?

(Windows) Template Studio - 🤷
Win2D - 🤷
Kiosk mode / assigned app - 🤷
Store for Business - 🤷

How do I even start/restart?

**Comment by pjmlp** [comment]
Nowadays I just follow the community calls to keep an eye on how the pain points are being sorted out, and if it is worthwhile to ever advise anyone to use WinUI for a greenfield application, other frameworks and Web, pay my bills now, after 6 years desilusion with WinRT management.

While the effort put by the team is very appreciated, I have the feeling that you lack resources, are lost in a kind of Microsoft civil war between MFC, ATL, Forms, WPF, UWP (bare bones), UWP/WinUI 2.0, Electron apps, WebWidget, Blazor Hybrid, PWAs, Android, WSLg, Azure making Windows irrelevant outside some key scenarios, while trying to keep the community motiviated to still care about WinUI 3.0 despite all its issues after 2 years in the making.

The team seems that still doesn't understand the API space and tooling that WinUI 3.0 is missing from WPF, boilerplate stuff like WinRT.Interop is sold as not a big deal, stuff gets announced like CsWinRT, C++/WinRT and then the tooling doesn't follow, there are only a couple of blog posts on Windows Developer introducing stuff that doesn't get mentioned again.

It is great that C++/WinRT lack of tooling is now being addressed, however who had the clever idea to introduce it, deprecate C++/CX about 4 years ago, without ever thinking about how clunky its developer experience is? Is the way that we paying customers should be treated?

Then we have a roadmap that in practice means the community has to hope for the next two years things don't change, like we had to hope we would get XAML Islands, Reunion, Windows 10X and then management decided otherwise.

In what concerns my point of view, the community calls only make sense if they are to actually dive into the issues, and how they are going to be actually sorted out, not selling dreams if we still burn more of our time in the next 2 years waiting for WinUI 3.0 to reach feature parity with existing frameworks.

I understand it is hard and the team probably has the hands tied in this process, however too many bridges have been burned since Windows 8, and the community can't be waiting forever.

So keep the community calls if it is to actually dive into the issues with detailed feedback, if it is to keep it high level, selling promises that stuff will come but you cannot talk about it right now, then no need for community calls, we have the Windows Developer blog announcements for that.

**Comment by dongle-the-gadget** [comment]
> who had the clever idea to introduce it

Kenny Kerr, but anyway I'm not sure who decided to deprecate C++/CX in favor of C++/WinRT when the Visual Studio team haven't made the tooling perfect.

**Comment by dongle-the-gadget** [comment]
Win2D: `Microsoft.Graphics.Win2D` NuGet package?

**Comment by wbokkers** [comment]
I think C++/WinRT was the right move. But it can only work with powerful tooling. Now Kenny Kerr is working on Azure IoT. Azure gets all the attention at Microsoft. Maybe WinUI should be rebranded as Azure UI.

**Comment by dongle-the-gadget** [comment]
> I think C++/WinRT was the right move. But it can only work with powerful tooling. Now Kenny Kerr is working on Azure IoT. Azure gets all the attention at Microsoft. Maybe WinUI should be rebranded as Azure UI.

Honestly I just want MS to continue working on C++/CX. I really loathe C++/WinRT after so many attempts to get it working.

**Comment by mrlacey** [comment]
@myokeeh I expect a Template Studio announcement next month 😉

**Comment by MEK3DK** [comment]
I think this post should be renamed to: "Let us hear about your frustrations regarding WinUI" and then create a new one regarding call feedback. :-)

I think this post have gone to much off track to be usefull for feedback.

@gabbybilka please fix the post.

**Comment by wbokkers** [comment]
An attempt to answer the questions:

**What should we do more of?**

* Answer questions listed in the feedback on a more technical level when possible. 
* Show progress on the github issues.
* Show changes in prioritization on the github issues.
* Try to meet the community where they are struggling getting things to work: migration paths, examples, workarounds,... 
* Really listen to feedback. Try to understand frustration.

**What should we stop doing/do less of?**

* Market/highlight new features or products before there is a stable product

**Should we make changes to the format of the call? E.g. more formal/informal, more time for a specific content area, hosted on a different platform?**

* Keep it informal. Don't mix in marketing stuff.

**At a general level, we're also interested in what kinds of demos & spotlights you most appreciate.** 

* How-to guides. Keep them simple. Provide downloadable examples.

**Comment by Rarisma** [comment]
Please improve the quality of the calls, considering this is a business and not some guy in thier basement doing a passion project. I'm pretty sure MS should be putting thier best foot forward here.

The calls only seen by those who have a deep intrest in winui so please stop trying to push marketing stuff aswell, as it stands the video has just under 1k views.

The samples repo is not very helpful and in my opinion should be rolled into the xaml controls libary.
I am extremely disapointed in the fact that the team is not pushing for open source, it should be the number one priorty as it would increase devlopment speed and adoption rates as winui can be frustating at times to develop for.

**Comment by Gavin-Williams** [comment]
More talk about graphics, interop and tooling around interop. There is still no good solution to calling DirectX from C#. My graphics work has mostly stopped for the past 2 years, because the new C++ -> C++/WinRT -> C#/WinRT -> C# system is just unusable. I have not in all this time been able to get it to work. I've tried several times. It's complex, it's fragile, it's cumbersome.

Also, can we see some demo's of calling Direct X using C#/WinRT / C++/WinRT. And give us some benchmarks please comparing:
  (1) SharpDX 
  (2) C++/CX 
  (3) CsWinRT+CppWinRT
  
Draw-call benchmarks, and look at thread-freezing due to GC. And also some comparisons of difficulty in the 3 approaches and a considerate assessment of whether it is realistic to expect people to use option 3. And what is the uptake of option 3? Are you seeing graphics projects using option 3 in the same numbers we saw people using SharpDX and C++/CX (XNA could be included here as well) or have you basically killed DirectX interop overnight with your new system?

And I am aware of TerraFX, which is promising in some ways, but for me, it makes a huge mistake by not C#'ifying it's API's, so it looks like some horrible mashup of low level C#, C++ and the poorly designed DX API signatures.

**Comment by Balkoth** [comment]
I just watched the last community call, this are my takeaway points:

- I am shocked about how bad the quality of the presentation (audio, video) is.
- The audio is constantly lagging behind the voice translation, so that it is almost unusable.
- Not your whole audience are native english speakers, so brief your guests that it is important to speak at a reasonable pace and with clear articulation.
- This maybe a personal thing, but kitchen streams from a fortune 500 company look cringe und unprofessional.

**Comment by lhak** [comment]
I would definitely like to see if feedback for specific bugs is given (what is the status, when it is going to be fixed). Currently, as others have also observed, serious issues without workarounds are ignored, sometimes for years.

**Comment by VigneshVoid** [comment]
Hi everyone,
We have migrated our app from UWP to Windows App SDK with WinUI 3. While migrating we faced a quite few challenges and had few inferences which we would like to share here. 


**Challenges faced:**
1. Finding an alternative to Window.Current, Dispatcher, ApplicationView.GetForCurrentView and implementing them.
2. We have used FocusManager for accessing fields in form with Arrow keys. Implementation related to FocusManager was quite intriguing as FindPreviousFocusableElement and FindNextFocusbleElement is not supported on Windows App SDK.
3. Our old UWP app had multi-window support and changing it to single window app was bit engaging. However multi-window support is available in WIndows App SDK 1.0.1 now and we will adapt the same.
4. Share UI replacement is still challenging as it is not working properly in some of the systems. [Reference](https://github.com/microsoft/WindowsAppSDK/issues/2174)
5. The significant part of the migration was the UI and style changes made in line with WinUI 3. And supporting WIndows 11 changes was on top of that.
6. Packaging was another bit complex thing to do. We initially tried Single Project MSIX Packaging but due to some bugs it was not working. So we achieved packaging through WAP Project.
7. Since Web Authentication Broker was not supported by WIndows App SDK, we had to use WebView2 for authentication.
8. We faced an issue in WebView2 with select field which is now fixed in Windows App SDK 1.0.1 update. Thanks to @marb2000  for acknowledging it.
9. We have used EmailManager for mailing purpose. Since it is not supported in Windows App SDK we have used Share UI for showing the mail option.
10. We have used OneDriveClient for our UWP app for accessing OneDrive. Since it is not available in Windows App SDK, we switched to Graph API to support OneDrive. But the authentication is done for every app session which needs to be addressed.


**Inferences:**
1. The major thing we miss in Windows App SDK is the unhandled exception handling which was very much useful in UWP app. [Reference](https://github.com/microsoft/WindowsAppSDK/discussions/1915)
2. The exceptions thrown in Windows App SDK app are not more meaningful and it is very difficult to analyze its root cause.
3.  AppWindow.TitleBar api is only available for Windows 11 so it is bit complex for customizing the Titlebar for Windows 10. [Reference](https://github.com/microsoft/microsoft-ui-xaml/issues/6308)


**Note:**
You can contact us for further discussions or demo session and we are pleased to engage with you.

**Comment by dongle-the-gadget** [comment]
For window related stuff, you might wanna try `AppWindow`, and for `Dispatcher`, use `DispatcherQueue` instead.

**Comment by VigneshVoid** [comment]
> Demos of applications using WinUI & their development journey

Hi @gabbybilka ,
As you mentioned that you're interested in giving a demo of applications using WinUI and development journey, we would like to be a part of it. We already shared our development journey in the previous comment and sharing here the [application link](https://www.microsoft.com/store/apps/9P1K7PZCTMS0?cid=WinuiCommunityCall)

If you want to check out our application we can provide you with the testing credentials. Kindly reach out to us for further discussion.

**Comment by MEK3DK** [comment]
**What is bad/suggestions:**

- Most important thing: Separate calls for WinUI 2 and WinUI 3, or separate sections.
  Why: Now that you are sticking to these almost identical names for 2 different things you need to make it easy to separate information. In some previous calls you have had guest present stuff and you are not sure what they are talking about because they just say WinUI. 
  
- Fix the sound/video quality of the calls. It is embarrassing that Microsoft shows something with such a poor connection. The last one was especially bad, but it has often been a problem.

- As you say it is 3+ years now, so the intro every time about what WinUI is should not be needed, just show a link if people want to know more. QA time much more important.

- Keep marketing stuff out, for now this is about the technical stuff.

- Have some update on the bug/Github issues situation. Currently we just know there is massive amounts of bugs, is it going up/down, how many are critical and so on.

- Showing actual code examples is always nice, but you need to record them somehow and make them available separately. Atm there is a lack of WinUI 3 examples/documentation (natural for a somewhat new tech).

- If possible, host it on something else then YouTube. Chat system is not suited for these kinds of calls. However other things are more important.

- In many calls you have stated that you would answer the missed questions in the forum at a later point in time, but often that has not happened. Some because you likely did not want to answer and some because you did not have any information. Answer all the community call questions in the forum and if there is something you don’t want to answer atm just be honest and say so.

**What is good:** 

- The community calls themselves. A very good way to interact directly with the community.
- The people from Microsoft at the call seems very capable of answering the different questions.
- You keep calm and answers question even when they are a bit harsh.
- The informal setting.

Unrelated: Overall a community manager is needed.

Finally: Please respond/act on this thread when time permits, much other stuff here has gone into oblivion.

**Comment by mdtauk** [comment]
It would be good (maybe not every call) that you discuss the future plans, current progress, and the current thinking on various features and controls.

Discussion on where people feel WinUI is lacking, to help prioritise things.

Examples and case studies where Microsoft is currently using WinUI 3 in it's products.

**Comment by gabbybilka** [comment]
> Finally: Please respond/act on this thread when time permits, much other stuff here has gone into oblivion.

I really appreciate all of the feedback around the community call format & content - it's more than I expected - thank you! I'm taking some of the top/common comments to improve our next call in April or keeping them on deck for a future call, including but not limited to: upping the stream & production quality, different content quality/type (more technical walkthroughs into existing features, insight into the closed source code, MSFT apps using WinUI 3, community spotlights/blogs) and reducing emphasis on upcoming features in unreleased versions.

For some of the feedback less/unrelated to the call itself, I've brought that to our team (e.g. docs improvements, bug status visibility & action) and if there's news to share, will report on that at the call.

After the April 20th call, let us know what you thought about the changes. I'll leave this issue open for then.

**Comment by gabbybilka** [comment]
Hey all - how was the April 20th community call? Any subjects you're interested in learning more about in a similar format to how we presented HybridCRT? Thanks for joining!

Recording [here](https://www.youtube.com/watch?v=bNHGU6xmUzE) (we'll be sure to post the correct link next time!)

**Comment by kmgallahan** [comment]
@gabbybilka 

Good call except the first 15 mins of sitting waiting for the stream to start in the wrong place... 😅

First some things to reiterate from my first comment above:

* How is rendering & composition done behind the scenes with WinUI, as compared to UWP XAML & WPF
* Provide a list of internally suggested topics from the team that you guys would actually like to talk about that the community can vote on a week or so in advance
* Make a decision that you'd be willing to dedicate some developer time each month to service a UI control or other area, present some options to the community which ones you'd like to tackle in the near term before the call for voting, present your plan to tackle it (some GitHub issues) during the call, then immediate feedback and discussion can be had during the Q&A
* Talk about the structure / architecture / quantity of any of the non-open source code

In addition:

* Present an overview of the team members, their WinUI specialties, and experience doing Windows UI development - we see you all in the call, but for the most part have no idea who does what and if there is anyone else behind the scenes. Would also be great to see how the WinUI team & WASDK team relate (or if the same thing), and where they fit into the greater MS org chart. All this is to get a better understand of who is working on things, what they do, and how far they reside from other people/teams in MS (and hopefully give people pause before lashing out at "M$" for not having infinite developers and a perfect UI framework.
* Explain why the debugging experience feels so bad as I mentioned [here](https://github.com/microsoft/microsoft-ui-xaml/issues/6980#issuecomment-1104078087) when compared to UWP / WPF / a normal .NET Core project. What is going on behind my C# code they makes it impossible to set a break point while an app is running most of the time....
* Consider presenting time & effort estimates for some of the most requested features from the community so that they can get a better idea of why some things aren't a priority right now. Examples: people really want a designer, advanced C++ datagrid, and acrylic / mica / transparent backgrounds / context menus presented outside the application window. The idea is to communicate that some things are wayyy more complicated & time consuming to work on than people think.
* Related to the above, what are all the "areas" of WinUI/WASDK that the team organizes work around in your internal issue tracking systems? In this repo we can view the available issue tags to get an idea of this, but a visual giving a bigger picture view of all the areas would be interesting.

**Comment by lukeblevins** [comment]
In a community call, we want to hear progress about what roadmap items were accomplished before they're released.

I think the idea to "reduce emphasis on features from unreleased WinAppSDK versions" is a _really_ bad one.

**Comment by deanchalk** [comment]
I'm looking to migrate several large enterprise WPF apps to something else around the later part of 2023. I have been interested in WinUI for a couple of years and have listened to several community calls.
For me, they would be improved by:

- Drop the ubiquitous intro, it's a bit tedious. If a listener doesn't know what the community call is for they shouldn't be there
- Forget marketing and announcements about future features (or keep to a bare minimum) - we need to concentrate on getting the current features working
- Focus on devs that have some experience rather than worrying about noobs - and reducing everything to the lowest common denominator
- drop the interactive Q&A - the questions getting answered are usually not that interesting. Why not search through this repo and discussions and find the big pain points and talk about those - and how your going to fix them
- I need regular updates on the progress to open source WinUI3 - I can't invest in a buggy technology unless I can delve into the code and either suggest a fix or create my own workaround. Getting to know the source code will help a lot in understanding WinUI3's weaknesses and strengths - and using that knowledge to make better app design decisions
- make it longer, with more technical deep-dives - especially around c++/Winrt use-cases

Thanks

**Reply by Gavin-Williams** [reply]
If they dropped the interactive Q&A, customer engagement would drop IMO, I think it's one of the drawcards, that people can raise issues or find out about something. Why not have both an interactive Q&A and some time to address repo issues? People are always asking for more feedback from MS, and here you are telling them to reduce it, I honestly don't get the logic. There aren't many places we can ask product teams questions in real time.

**Reply by pjmlp** [reply]
To be fair, it already dropped for all of us that had enough with all the rewrites that were pushed into since Windows 8 was introduced, and aren't willing to wait for Windows 12 to get back all UWP features, including VS tooling.
